File: scripts/setup-schema.py
#!/usr/bin/env python3
"""
Execute Supabase schema SQL directly using PostgreSQL connection
"""

import os
import sys
import re
from pathlib import Path

# Try to import psycopg2
try:
    import psycopg2
except ImportError:
    print("❌ psycopg2 not found. Installing...")
    os.system("pip install psycopg2-binary python-dotenv")
    import psycopg2

from dotenv import load_dotenv

# Load environment variables
load_dotenv('.env.local')

SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
SUPABASE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')

if not SUPABASE_URL:
    print("❌ Missing NEXT_PUBLIC_SUPABASE_URL in .env.local")
    sys.exit(1)

# Extract PostgreSQL connection details from Supabase URL
# URL format: https://project-id.supabase.co
match = re.match(r'https://([^.]+)\.supabase\.co', SUPABASE_URL)
if not match:
    print("❌ Invalid SUPABASE_URL format")
    sys.exit(1)

project_id = match.group(1)

# Supabase PostgreSQL connection details
DB_HOST = f"{project_id}.db.supabase.co"
DB_PORT = 5432
DB_NAME = "postgres"
DB_USER = "postgres"
DB_PASSWORD = os.getenv('SUPABASE_DB_PASSWORD')

# If password not in env, ask user or use service role key
if not DB_PASSWORD:
    print("⚠️  SUPABASE_DB_PASSWORD not found in .env.local")
    print("    You need the database password. Options:")
    print("    1. Add it to .env.local: SUPABASE_DB_PASSWORD=<password>")
    print("    2. Or get it from Supabase Dashboard → Settings → Database")
    DB_PASSWORD = input("\n   Enter PostgreSQL password: ")
    
    if not DB_PASSWORD:
        print("❌ Password required")
        sys.exit(1)

# Read schema SQL from file
schema_file = Path(__file__).parent / "supabase-schema.sql"
if not schema_file.exists():
    print(f"❌ Schema file not found: {schema_file}")
    sys.exit(1)

with open(schema_file) as f:
    schema_sql = f.read()

print("═" * 60)
print("📋 SUPABASE SCHEMA CREATION")
print("═" * 60)
print(f"\n🔗 Connecting to: {DB_HOST}\n")

try:
    # Connect to PostgreSQL
    conn = psycopg2.connect(
        host=DB_HOST,
        port=DB_PORT,
        database=DB_NAME,
        user=DB_USER,
        password=DB_PASSWORD,
        sslmode='require'
    )
    
    cursor = conn.cursor()
    
    print("✅ Connected to PostgreSQL\n")
    print("🔅 Executing schema SQL...\n")
    
    # Execute the schema SQL
    cursor.execute(schema_sql)
    conn.commit()
    
    print("✅ Schema created successfully!\n")
    
    # Verify tables were created
    cursor.execute("""
        SELECT table_name FROM information_schema.tables 
        WHERE table_schema = 'public'
        ORDER BY table_name
    """)
    
    tables = cursor.fetchall()
    print("📊 Tables created:")
    for table in tables:
        print(f"   ✓ {table[0]}")
    
    print("\n✅ SCHEMA SETUP COMPLETED!\n")
    
    cursor.close()
    conn.close()
    
except psycopg2.OperationalError as e:
    print(f"❌ Connection failed: {e}")
    print("\n   Troubleshooting:")
    print("   1. Check your database password is correct")
    print("   2. Ensure you have network access to Supabase")
    print("   3. Check that the project ID is correct")
    sys.exit(1)
except psycopg2.Error as e:
    print(f"❌ Database error: {e}")
    print(f"\n   Details: {e.diag.message_primary if hasattr(e, 'diag') else str(e)}")
    sys.exit(1)
except Exception as e:
    print(f"❌ Unexpected error: {e}")
    sys.exit(1)

File: scripts/setup-schema.js
#!/usr/bin/env node
/**
 * Execute Supabase schema SQL directly using PostgreSQL connection
 * 
 * This script connects to Supabase PostgreSQL database and executes the schema SQL
 * 
 * Requirements:
 * - Database password from Supabase Dashboard → Settings → Database → Database password
 * - Add to .env.local: SUPABASE_DB_PASSWORD=<password>
 * 
 * Or you can execute manually:
 * 1. Go to https://app.supabase.com
 * 2. Project → SQL Editor → New Query
 * 3. Copy content from scripts/supabase-schema.sql
 * 4. Click Run
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import dotenv from 'dotenv';

dotenv.config({ path: '.env.local' });

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!;
const schemaFile = path.resolve(path.dirname(fileURLToPath(import.meta.url)), './supabase-schema.sql');

if (!supabaseUrl) {
  console.error('❌ Missing NEXT_PUBLIC_SUPABASE_URL in .env.local');
  process.exit(1);
}

const urlMatch = supabaseUrl.match(/https:\/\/([^.]+)\.supabase\.co/);
if (!urlMatch) {
  console.error('❌ Invalid SUPABASE_URL format');
  process.exit(1);
}

const projectId = urlMatch[1];

console.log('═══════════════════════════════════════════════════════════');
console.log('📋 SUPABASE SCHEMA SETUP');
console.log('═══════════════════════════════════════════════════════════\n');

console.log('To create the database schema, you have options:\n');
console.log('OPTION 1: Manual Setup (Recommended - no password needed)');
console.log('────────────────────────────────────────────────────────');
console.log('1. Go to https://app.supabase.com');
console.log('2. Select your project');
console.log('3. Navigate to: SQL Editor → New Query');
console.log('4. Copy ALL content from: scripts/supabase-schema.sql');
console.log('5. Paste into the SQL editor and click "Run"\n');

console.log('OPTION 2: Automatic Setup (requires database password)');
console.log('────────────────────────────────────────────────────────');
console.log('1. Get your database password from:');
console.log('   Supabase Dashboard → Settings → Database → Database password');
console.log('2. Add to .env.local:');
console.log('   SUPABASE_DB_PASSWORD=<your_password>');
console.log('3. Run: npm run db:setup-schema\n');

console.log('Schema file location: ' + schemaFile);
console.log('Schema SQL preview (first 500 chars):');
const sql = fs.readFileSync(schemaFile, 'utf-8');
console.log('───────────────────────────────────────────────────────');
console.log(sql.substring(0, 500));
console.log('...\n');

const dbPassword = process.env.SUPABASE_DB_PASSWORD;
if (!dbPassword) {
  console.log('⏳ No SUPABASE_DB_PASSWORD found in .env.local');
  console.log('   Use Option 1 (Manual Setup) above, or provide the password.\n');
  console.log('📖 After creating the schema, run:');
  console.log('   npm run db:import\n');
  process.exit(0);
}

console.log('🔐 Found SUPABASE_DB_PASSWORD, attempting automatic setup...\n');

// Try to use pg library if available
try {
  const pg = await import('pg');
  const { Client } = pg.default;

  const client = new Client({
    host: `${projectId}.db.supabase.co`,
    port: 5432,
    database: 'postgres',
    user: 'postgres',
    password: dbPassword,
    ssl: { rejectUnauthorized: false },
  });

  console.log(`🔗 Connecting to ${projectId}.db.supabase.co...\n`);

  try {
    await client.connect();
    console.log('✅ Connected to PostgreSQL\n');

    console.log('🔅 Executing schema SQL...\n');
    await client.query(sql);

    console.log('✅ Schema created successfully!\n');

    // Verify tables
    const result = await client.query(`
      SELECT table_name FROM information_schema.tables 
      WHERE table_schema = 'public'
      ORDER BY table_name
    `);

    console.log('📊 Tables created:');
    for (const row of result.rows) {
      console.log(`   ✓ ${row.table_name}`);
    }

    console.log('\n✅ SCHEMA SETUP COMPLETED!\n');
    console.log('📖 Next, run: npm run db:import\n');

    await client.end();
  } catch (err: any) {
    console.error('❌ Error:', err.message);
    if (err.message.includes('password')) {
      console.log('\n⚠️  Password authentication failed.');
      console.log('   Check your SUPABASE_DB_PASSWORD is correct.');
    }
    process.exit(1);
  }
} catch (err: any) {
  console.log('⚠️  pg library not installed. Installing...\n');
  console.log('For automatic schema setup, install the pg package:');
  console.log('   npm install pg\n');
  console.log('Or use the manual setup option above.\n');
  process.exit(1);
}
